### PR TITLE
some contract v2.3 tuning

### PR DESCRIPTION
**Pull request by shileiwill**
As title.

### CODE REVIEW

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol [hunk 1]
```diff
@@ -73,6 +76,7 @@ contract AutomationRegistrar2_3 is TypeAndVersionInterface, ConfirmedOwner, IERC
 
   struct PendingRequest {
     address admin;
+    IERC20 billingToken;
     uint96 balance;
   }
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:81]: I think we could get rid of this entire struct and change `s_pendingRequests` to...
```solidity
mapping(bytes32 => PendingRequest) private s_pendingRequests;
```
... this is what our friends recommended too :)

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol [hunk 2]
```diff
@@ -217,22 +220,22 @@ contract AutomationRegistrar2_3 is TypeAndVersionInterface, ConfirmedOwner, IERC
...
     if (request.admin == address(0)) {
       revert RequestNotFound();
     }
     delete s_pendingRequests[hash];
-    bool success = i_LINK.transfer(request.admin, request.balance);
-    if (!success) {
-      revert TransferFailed(request.admin);
-    }
+
+    request.billingToken.safeTransfer(request.admin, request.balance);
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:237]: 👍

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol [hunk 3]
```diff
@@ -375,8 +381,11 @@ contract AutomationRegistrar2_3 is TypeAndVersionInterface, ConfirmedOwner, IERC
       s_triggerRegistrations[params.triggerType].approvedCount++;
       upkeepId = _approve(params, hash);
     } else {
-      uint96 newBalance = s_pendingRequests[hash].balance + params.amount;
-      s_pendingRequests[hash] = PendingRequest({admin: params.adminAddress, balance: newBalance});
+      s_pendingRequests[hash] = PendingRequest({
+        admin: params.adminAddress,
+        billingToken: params.billingToken,
+        balance: params.amount
+      });
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:388]: I think here we could just do `s_pendingRequests[hash] = true;`

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol [hunk 4]
```diff
@@ -10,7 +10,7 @@ import {Chainable} from "../../Chainable.sol";
 import {IERC677Receiver} from "../../../shared/interfaces/IERC677Receiver.sol";
 import {OCR2Abstract} from "../../../shared/ocr2/OCR2Abstract.sol";
 import {IERC20} from "../../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/IERC20.sol";
-import {SafeCast} from "../../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/math/SafeCast.sol";
+import {SafeERC20} from "../../../vendor/openzeppelin-solidity/v4.8.3/contracts/token/ERC20/utils/SafeERC20.sol";
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol:13]: It doesn't look like we use any library functions in this contract? So I think all this really did is increase the contract size. Not a strong opinion, but I might have left this contract as-is to keep the hot path cheaper.

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol [hunk 5]
```diff
@@ -201,6 +207,7 @@ contract AutomationRegistryLogicA2_3 is AutomationRegistryBase2_3, Chainable {
         billingToken.safeTransfer(destination, balanceToTransfer);
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol:207]: could we move this transfer to outside the loop and drop the `if` clause? I think that would look cleaner.

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol [hunk 6]
```diff
@@ -356,6 +358,10 @@ contract AutomationRegistrar2_3 is TypeAndVersionInterface, ConfirmedOwner, IERC
     }
     bytes32 hash = keccak256(abi.encode(params));
 
+    if (s_pendingRequests[hash].admin != address(0)) {
+      revert DuplicateEntry();
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:362]: can we add a foundry test for the duplicate registration scenario?

File: contracts/src/v0.8/automation/dev/test/AutomationRegistry2_3.t.sol [hunk 7]
```diff
@@ -1503,8 +1534,11 @@ contract MigrateReceive is SetUp {
       prevReserveBalances[2] - registry.getReserveAmount(address(weth))
     );
 
-    // assert token have been transfered
-    assertEq(linkToken.balanceOf(address(newRegistry)), newRegistry.getBalance(linkUpkeepID));
+    // assert token have been transferred
+    assertEq(
+      linkToken.balanceOf(address(newRegistry)),
+      newRegistry.getBalance(linkUpkeepID) + newRegistry.getBalance(linkUpkeepID2)
+    );
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/test/AutomationRegistry2_3.t.sol:1541]: awesome 🙌

**Review by RyanRHall — CHANGES_REQUESTED** [review]
awesome job 🎉

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol (hunk 1 above)
Comment (shileiwill) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:81]: hmm, can you elaborate a bit more on this? if we dont have this PendingRequest struct, when users cancel their pending upkeep, how we do refund?

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol (hunk 3 above)
Comment (shileiwill) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:388]: ditto, would like some clarification.

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol (hunk 4 above)
Comment (shileiwill) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol:13]: yes, my bad, i forgot to clean this up after moving the addFund

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol (hunk 5 above)
Comment (shileiwill) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol:207]: yes, agreed.

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol (hunk 6 above)
Comment (shileiwill) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:362]: will do.

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol (hunk 1 above)
Comment (RyanRHall) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistrar2_3.sol:81]: actually nevermind - forget I said anything 🙃

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol [hunk 8]
```diff
@@ -228,37 +227,6 @@ contract AutomationRegistry2_3 is AutomationRegistryBase2_3, OCR2Abstract, Chain
...
-      // ERC20 payment
-      bool success = upkeep.billingToken.transferFrom(msg.sender, address(this), amount);
-      if (!success) revert TransferFailed();
-    } else {
-      // native payment
-      i_wrappedNativeToken.deposit{value: amount}();
-    }
-
-    emit FundsAdded(id, msg.sender, amount);
-  }
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol:260]: could we add this back to the root contract now?

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol [hunk 9]
```diff
@@ -163,12 +164,13 @@ contract AutomationRegistryLogicA2_3 is AutomationRegistryBase2_3, Chainable {
...
     bytes[] memory triggerConfigs = new bytes[](ids.length);
     bytes[] memory offchainConfigs = new bytes[](ids.length);
+
     for (uint256 idx = 0; idx < ids.length; idx++) {
       id = ids[idx];
       upkeep = s_upkeep[id];
 
       if (idx == 0) {
-        billingToken = s_upkeep[id].billingToken;
+        billingToken = upkeep.billingToken;
```
Comment (RyanRHall) [inline contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryLogicA2_3.sol:173]: 👍

**Review by RyanRHall — COMMENTED** [review]
Just the one question!

File: contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol (hunk 8 above)
Comment (shileiwill) [reply contracts/src/v0.8/automation/dev/v2_3/AutomationRegistry2_3.sol:260]: unfortunately no~ the addFunds requires SafeCast and SafeERC20 libraries, so when we move back the function, we also need to move back the libs. it is oversize.

**Review by RyanRHall — APPROVED** [review]
